File: macos/BlindfoldCore/Sources/BlindfoldCore/StatusClient.swift
import Foundation

/// Errors the `/v1/status` client can fail closed on.
public enum StatusClientError: Error, Equatable, Sendable {
    /// The supplied base URL does not resolve to the loopback interface — refused
    /// rather than silently polling somewhere other than the local proxy
    /// (ADR-0040's egress-discipline clause: this core only ever calls loopback).
    case nonLoopbackBaseURL
}

/// The network boundary `StatusClient` calls through — stubbed in tests
/// (leak-audit's seam-stub pattern), backed by `URLSession` in the real app.
public protocol StatusFetching: Sendable {
    func fetchStatus(from url: URL) async throws -> Data
}

/// The cadence timer `pollLoop` waits on — stubbed in tests so cadence is asserted
/// on recorded intervals rather than real wall-clock waits.
public protocol Sleeping: Sendable {
    func sleep(seconds: Double) async throws
}

/// `/v1/status`'s payload, decoded to only the fields the state machine needs
/// (`state`, plus #180's `unprotected_mode` alarm fields) — never the full
/// scrubbed-but-broader contract (`dependencies`, `blocks`, ...), so this core
/// never even holds a field it doesn't reduce over.
public struct StatusPayload: Decodable, Equatable, Sendable {
    /// ADR-0038's bounded override: overlays the Unprotected alarm on top of the
    /// five-state machine, read verbatim from #180's `/v1/status` addition.
    public struct UnprotectedMode: Decodable, Equatable, Sendable {
        public let active: Bool
        public let bound: String?
        public let remainingSeconds: Double?

        public init(active: Bool, bound: String?, remainingSeconds: Double?) {
            self.active = active
            self.bound = bound
            self.remainingSeconds = remainingSeconds
        }

        enum CodingKeys: String, CodingKey {
            case active, bound
            case remainingSeconds = "remaining_seconds"
        }
    }

    public let state: String
    public let unprotectedMode: UnprotectedMode?

    public init(state: String, unprotectedMode: UnprotectedMode? = nil) {
        self.state = state
        self.unprotectedMode = unprotectedMode
    }

    enum CodingKeys: String, CodingKey {
        case state
        case unprotectedMode = "unprotected_mode"
    }
}

/// Polls the local proxy's `/v1/status` on a short cadence (ADR-0039/0040).
///
/// Egress discipline: constructing a client against anything but the loopback
/// interface fails closed at init, before any request can be made.
public final class StatusClient: Sendable {
    public let baseURL: URL
    private let fetcher: StatusFetching

    public init(baseURL: URL, fetcher: StatusFetching) throws {
        guard StatusClient.isLoopback(baseURL) else {
            throw StatusClientError.nonLoopbackBaseURL
        }
        self.baseURL = baseURL
        self.fetcher = fetcher
    }

    public func poll() async throws -> StatusPayload {
        let data = try await fetcher.fetchStatus(from: baseURL)
        return try JSONDecoder().decode(StatusPayload.self, from: data)
    }

    /// Polls on a short cadence (ADR-0039). `iterations` bounds the loop for
    /// deterministic testing; `nil` polls until the enclosing `Task` is cancelled.
    public func pollLoop(
        intervalSeconds: Double,
        sleeper: Sleeping,
        iterations: Int? = nil,
        onUpdate: (StatusPayload) -> Void
    ) async throws {
        var remaining = iterations
        while !Task.isCancelled {
            if let remaining, remaining <= 0 { break }
            let payload = try await poll()
            onUpdate(payload)
            remaining = remaining.map { $0 - 1 }
            if let remaining, remaining <= 0 { break }
            try await sleeper.sleep(seconds: intervalSeconds)
        }
    }

    static func isLoopback(_ url: URL) -> Bool {
        guard let host = url.host else { return false }
        return host == "127.0.0.1" || host == "localhost" || host == "::1"
    }
}

File: windows/Blindfold.Core/StatusClient.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Blindfold.Core;

/// <summary>
/// <c>/v1/status</c>'s payload, decoded to only the narrow contract the tray app reduces
/// over (issue #194): <c>state</c>, the ADR-0038 <c>unprotected_mode</c> alarm fields, and
/// a dependencies-down count. This core never holds a field it doesn't reduce over — no
/// dependency name, no scrubbed detail string, no entity value.
/// </summary>
public sealed class StatusPayload
{
    /// <summary>
    /// ADR-0038's bounded override: overlays the Unprotected alarm on top of the
    /// five-state machine, read verbatim from the proxy's <c>/v1/status</c> addition.
    /// </summary>
    public sealed class UnprotectedModeFields
    {
        public bool Active { get; init; }
        public string? Bound { get; init; }
        public double? RemainingSeconds { get; init; }
        /// <summary>
        /// Issue #197's submenu visibility gate (mirrors #187/#188's Swift
        /// <c>capabilityEnabled</c>): whether the operator has opted the capability into
        /// existence, read verbatim regardless of whether the mode is currently active.
        /// Fail-closed: an absent field reads as capability-disabled, never enabled.
        /// </summary>
        public bool CapabilityEnabled { get; init; }

        public UnprotectedModeFields(bool active, string? bound, double? remainingSeconds, bool capabilityEnabled = false)
        {
            Active = active;
            Bound = bound;
            RemainingSeconds = remainingSeconds;
            CapabilityEnabled = capabilityEnabled;
        }
    }

    public string State { get; }
    public UnprotectedModeFields? UnprotectedMode { get; }
    /// <summary>
    /// The header line's dependencies-down count — only the count crosses the boundary.
    /// <c>dependencies</c> itself (name -> healthy/detail/latency_ms) is decoded
    /// transiently in <see cref="Decode"/> and discarded, never stored.
    /// </summary>
    public int DependenciesDown { get; }
    /// <summary>
    /// Issue #197's review-inbox deep-link count, read straight from
    /// <c>review_inbox.pending</c> -- never inbox contents.
    /// </summary>
    public int ReviewInboxPending { get; }
    /// <summary>
    /// Issue #197's blocks deep-link count, read straight from <c>blocks.count</c> --
    /// never the <c>recent</c> block records themselves.
    /// </summary>
    public int BlocksCount { get; }
    /// <summary>
    /// Issue #197's "Finish setup →" visibility, read straight from <c>empty_store</c>.
    /// </summary>
    public bool EmptyStore { get; }

    public StatusPayload(
        string state,
        UnprotectedModeFields? unprotectedMode = null,
        int dependenciesDown = 0,
        int reviewInboxPending = 0,
        int blocksCount = 0,
        bool emptyStore = false)
    {
        State = state;
        UnprotectedMode = unprotectedMode;
        DependenciesDown = dependenciesDown;
        ReviewInboxPending = reviewInboxPending;
        BlocksCount = blocksCount;
        EmptyStore = emptyStore;
    }

    private sealed class UnprotectedModeWire
    {
        [JsonPropertyName("active")]
        public bool Active { get; set; }

        [JsonPropertyName("bound")]
        public string? Bound { get; set; }

        [JsonPropertyName("remaining_seconds")]
        public double? RemainingSeconds { get; set; }

        [JsonPropertyName("capability_enabled")]
        public bool CapabilityEnabled { get; set; }
    }

    private sealed class DependencyHealthEntry
    {
        [JsonPropertyName("healthy")]
        public bool Healthy { get; set; }
    }

    private sealed class ReviewInboxWire
    {
        [JsonPropertyName("pending")]
        public int Pending { get; set; }
    }

    private sealed class BlocksWire
    {
        [JsonPropertyName("count")]
        public int Count { get; set; }
    }

    private sealed class Wire
    {
        [JsonPropertyName("state")]
        public string State { get; set; } = "";

        [JsonPropertyName("unprotected_mode")]
        public UnprotectedModeWire? UnprotectedMode { get; set; }

        [JsonPropertyName("dependencies")]
        public Dictionary<string, DependencyHealthEntry>? Dependencies { get; set; }

        [JsonPropertyName("review_inbox")]
        public ReviewInboxWire? ReviewInbox { get; set; }

        [JsonPropertyName("blocks")]
        public BlocksWire? Blocks { get; set; }

        [JsonPropertyName("empty_store")]
        public bool EmptyStore { get; set; }
    }

    internal static StatusPayload Decode(string json)
    {
        var wire = JsonSerializer.Deserialize<Wire>(json)
            ?? throw new StatusClientException("empty /v1/status body");
        var unprotectedMode = wire.UnprotectedMode is { } m
            ? new UnprotectedModeFields(m.Active, m.Bound, m.RemainingSeconds, m.CapabilityEnabled)
            : null;
        var dependenciesDown = wire.Dependencies?.Values.Count(d => !d.Healthy) ?? 0;
        return new StatusPayload(
            wire.State,
            unprotectedMode,
            dependenciesDown,
            wire.ReviewInbox?.Pending ?? 0,
            wire.Blocks?.Count ?? 0,
            wire.EmptyStore);
    }
}

/// <summary>
/// Errors the <c>/v1/status</c> client can fail closed on.
/// </summary>
public sealed class StatusClientException : Exception
{
    public StatusClientException(string message) : base(message)
    {
    }
}

/// <summary>
/// The network boundary <c>StatusClient</c> calls through — stubbed in tests
/// (leak-audit's seam-stub pattern), backed by <c>HttpClient</c> in the real tray app.
/// </summary>
public interface IStatusFetching
{
    Task<string> FetchStatusAsync(Uri url);
}

/// <summary>
/// Polls the local proxy's <c>/v1/status</c> on a short cadence (ADR-0039/0040/0041).
///
/// Egress discipline: constructing a client against anything but the loopback interface
/// fails closed at construction, before any request can be made.
/// </summary>
public sealed class StatusClient
{
    public Uri BaseUrl { get; }
    private readonly IStatusFetching _fetcher;

    public StatusClient(string baseUrl, IStatusFetching fetcher)
    {
        if (!Uri.TryCreate(baseUrl, UriKind.Absolute, out var uri) || !IsLoopback(uri))
        {
            throw new StatusClientException("non-loopback base URL");
        }

        BaseUrl = uri;
        _fetcher = fetcher;
    }

    public async Task<StatusPayload> PollAsync()
    {
        var json = await _fetcher.FetchStatusAsync(BaseUrl);
        return StatusPayload.Decode(json);
    }

    internal static bool IsLoopback(Uri url)
    {
        // IdnHost strips the IPv6 bracket syntax Host retains (e.g. "[::1]" -> "::1"),
        // matching Foundation's URL.host on the Swift side.
        var host = url.IdnHost;
        return host is "127.0.0.1" or "localhost" or "::1";
    }
}
